Move SUBSCRIBE, PUBLISH, FETCH, TRACK_STATUS, and PUBLISH_NAMESPACE to bidi streams (#1389)
moq-wg/moq-transport: draft-ietf-moq-transport.md

The above list of conditions is not considered exhaustive.

When a subscriber detects a Malformed Track, it MUST send unsubsribe any

> subscription and cancel any fetch for that Track from that publisher by sending

Reviewer: EPARSE.  it must cancel the corresponding subscription or fetch

> subscription and cancel any fetch for that Track from that publisher by sending

Author: ```suggestion
When a subscriber detects a Malformed Track, it MUST cancel any corresponding
subscription or fetches for that Track from that publisher by sending
```

The above list of conditions is not considered exhaustive.

When a subscriber detects a Malformed Track, it MUST cancel any corresponding
subscription or fetches for that Track from that publisher
(see {{request-cancellation}}), and SHOULD deliver an error to the application.
If a relay detects a Malformed Track, it MUST immediately terminate downstream
subscriptions with PUBLISH_DONE and reset any fetch streams with
Status Code `MALFORMED_TRACK`. Object(s) triggering Malformed Track status

> MUST NOT be cached.

Reviewer 2: is this part of this PR ? Can we move it to a different PR

Author: This is existing text, I minorly updated it to remove references to UNSUBSCRIBE and FETCH_CANCEL.

in which case the stream data SHOULD be buffered until the control stream arrives
and setup is complete. If an implementation does not want to buffer or if it is
certain the message type is not supported, it MAY reset other bidirectional streams
before the session and control stream are established.

The control stream MUST NOT be closed at the underlying transport layer during the
session's lifetime.  Doing so results in the session being closed as a
`PROTOCOL_VIOLATION`.

### Request Cancellation and Rejection {#request-cancellation}

Once a request stream has been opened, the request MAY be cancelled by either
endpoint. Subscribers cancel requests if the response is no longer of interest;
Publishers cancel requests if they are unable to or choose not to respond.

Implementations SHOULD cancel requests by abruptly terminating any directions of a
stream that are still open. To do so, an implementation resets the sending parts of
streams and aborts reading on the receiving parts of streams; see ({{QUIC, Section 2.4}}).

When the Publisher rejects a request without performing any application processing,

> The PUBLISHER SHOULD send a REQUEST_ERROR and FINs the stream.

Reviewer: ```suggestion
The PUBLISHER SHOULD send a REQUEST_ERROR and FIN the stream.
```

streams.

A unidirectional stream containing Objects or bidirectional stream(s) beginning
with a message other than CLIENT_SETUP could arrive prior to the control stream,
in which case the stream data SHOULD be buffered until the control stream arrives
and setup is complete. If an implementation does not want to buffer or if
the message type is not supported, it MAY reset such bidirectional streams
before the session and control stream are established.

The control stream MUST NOT be closed at the underlying transport layer during the
session's lifetime.  Doing so results in the session being closed as a
`PROTOCOL_VIOLATION`.

### Request Cancellation and Rejection {#request-cancellation}

Once a request stream has been opened, the request MAY be cancelled by either
endpoint. Senders cancel requests if the response is no longer of interest;
Receivers cancel requests if they are unable to or choose not to respond.

Implementations SHOULD cancel requests by abruptly terminating any directions of
a stream that are still open using RESET_STREAM / RESET_STREAM_AT or
STOP_SENDING.

When the Publisher rejects a request without performing any application processing,

> The PUBLISHER SHOULD send a REQUEST_ERROR and FIN the stream.

Reviewer 2: would this apply to both publisher and the subscriber ?

streams.

A unidirectional stream containing Objects or bidirectional stream(s) beginning
with a message other than CLIENT_SETUP could arrive prior to the control stream,
in which case the stream data SHOULD be buffered until the control stream arrives
and setup is complete. If an implementation does not want to buffer or if
the message type is not supported, it MAY reset such bidirectional streams
before the session and control streams are established.

The control stream MUST NOT be closed at the underlying transport layer during the
session's lifetime.  Doing so results in the session being closed as a
`PROTOCOL_VIOLATION`.

### Request Cancellation and Rejection {#request-cancellation}

Once a request stream has been opened, the request MAY be cancelled by either
endpoint. Senders cancel requests if the response is no longer of interest;
Receivers cancel requests if they are unable to or choose not to respond.

Implementations SHOULD cancel requests by abruptly terminating any directions of
a stream that are still open using RESET_STREAM / RESET_STREAM_AT or
STOP_SENDING.

When the Publisher rejects a request without performing any application processing,

> The PUBLISHER SHOULD send a REQUEST_ERROR and FIN the stream.

Author: ```suggestion
When an endpoint rejects a request without performing any application processing,
it SHOULD send a REQUEST_ERROR and FIN the stream.
```

  specification.

INVALID_REQUEST_ID (0x4):
: The endpoint received a Request ID with an incorrect least significant
  bit for the sender, or a duplicate Request ID. See {{request-id}}.

INVALID_REQUIRED_REQUEST_ID (0x7):
: The endpoint received a Required Request ID that does not have the same

>   parity as the Request ID, or is less than 0. See {{required-request-id}}.

Author: ```suggestion
: The endpoint received a Required Request ID Delta that results
in an invalid Request ID. See {{required-request-id}}.
```

be updated by the subscriber using REQUEST_UPDATE.  Either endpoint
can terminate an `Established` subscription, moving it to the
`Terminated` state.  The subscriber terminates a subscription in the
`Pending (Subscriber)` or `Established` states by cancelling the

> bidirectional stream. The publisher terminates a subscription in the

Author: ```suggestion
`Pending (Subscriber)` or `Established` states by sending STOP_SENDING.
The publisher terminates a subscription in the
```

              |            |             |       | REQUEST_UPDATE
              |            +-------------+ <-----+
              |                 |    |                  |

>               +-- Cancel Stream |    | PUBLISH_DONE ----+

Author: ```suggestion
              +--- STOP_SENDING |    | PUBLISH_DONE ----+
```

REQUEST_ERROR to the application to inform the search for additional
subscribers for a namespace, or to abandon the attempt to publish under this
namespace. This might be especially useful in upload or chat applications. A
subscriber MUST send exactly one REQUEST_OK or REQUEST_ERROR as the first
message on the bidi stream in response to a PUBLISH_NAMESPACE. The publisher
SHOULD close the session with a protocol error if it receives more than one.

> A PUBLISH_NAMESPACE is withdrawn by sending STOP_SENDING on the bidi stream,

Reviewer: Why is this not a RESET_STREAM / FIN on the write half?

REQUEST_ERROR to the application to inform the search for additional
subscribers for a namespace, or to abandon the attempt to publish under this
namespace. This might be especially useful in upload or chat applications. A
subscriber MUST send exactly one REQUEST_OK or REQUEST_ERROR as the first
message on the bidi stream in response to a PUBLISH_NAMESPACE. The publisher
SHOULD close the session with a protocol error if it receives more than one.

> A PUBLISH_NAMESPACE is withdrawn by cancelling the Request

Author: I think Request is better, in which case we should probably define the term as well?

though the expectation is that all subscriptions will be able to send some data.

Given the critical nature of control messages and their relatively

> small size, the control stream and bidi request streams SHOULD be prioritized

Reviewer 3: maybe rank them by subscriber priority?

REQUEST_ERROR to the application to inform the search for additional
subscribers for a namespace, or to abandon the attempt to publish under this
namespace. This might be especially useful in upload or chat applications. A
subscriber MUST send exactly one REQUEST_OK or REQUEST_ERROR as the first
message on the bidi stream in response to a PUBLISH_NAMESPACE. The publisher
SHOULD close the session with a protocol error if it receives more than one.

A PUBLISH_NAMESPACE is withdrawn by cancelling the request
(see {{request-cancellation}}), although it is not a protocol error for
the subscriber to send a SUBSCRIBE or FETCH message for a track in a
namespace after the namespace is withdrawn.

> A subscriber can cancel the request to revoke acceptance of a

Reviewer 2: ```suggestion
A subscriber can cancel the request (see {{request-cancellation}}) to revoke acceptance of a
```

in the near future or it wants to reserve some bandwidth for control messages.

Given the critical nature of control messages and their relatively
small size, the control stream SHOULD be prioritized highest, followed by the

> bidi request streams and then all subscribed Objects. Control streams MAY be

Author: ```suggestion
bidi request streams and then all subscribed Objects. Bidi request streams MAY be
```

There is no specified limit to the number of publishers of a Track Namespace or
Track.  An implementation can use mechanisms such as REQUEST_ERROR,

> STOP_SENDING or PUBLISH_NAMESPACE_CANCEL if it cannot

Author: ```suggestion
unsubscribing via STOP_SENDING or PUBLISH_NAMESPACE_CANCEL if it cannot
```

relay is managed and is application specific.

When a publisher wants to stop new subscriptions for a published namespace it
sends a STOP_SENDING on the bidi stream to withdraw the PUBLISH_NAMESPACE.

> A subscriber indicates it will no longer subcribe to Tracks in a namespace

Author: ```suggestion
A subscriber indicates it will no longer subscribe to Tracks in a namespace
```

The authorization and identification of the publisher depends on the way the
relay is managed and is application specific.

When a publisher wants to stop new subscriptions for a published namespace, it
cancels the request (see {{request-cancellation}}) to withdraw the PUBLISH_NAMESPACE.

> A subscriber also indicates it will no longer subscribe to Tracks in a namespace

Reviewer: ```suggestion
A subscriber indicates it will no longer subscribe to Tracks in a namespace
```

# Control Messages {#message}

MOQT uses bidirectional streams to exchange control messages, as

> defined in {{session-init}}.  Every single message on a control stream is

Reviewer: ```suggestion
defined in {{session-init}}.  Every message on a control or request stream is
```

|-------|-----------------------------------------------------|
| 0x18  | FETCH_OK ({{message-fetch-ok}})                     |
|-------|-----------------------------------------------------|
| 0xD   | TRACK_STATUS ({{message-track-status}})             |
|-------|-----------------------------------------------------|
| 0x6   | PUBLISH_NAMESPACE  ({{message-pub-ns}})             |
|-------|-----------------------------------------------------|
| 0x8   | NAMESPACE  ({{message-namespace}})                  |
|-------|-----------------------------------------------------|
| 0xE   | NAMESPACE_DONE  ({{message-namespace-done}})        |
|-------|-----------------------------------------------------|

> | 0xC   | PUBLISH_NAMESPACE_CANCEL ({{message-pub-ns-cancel}})|

Author: I agree, but currently it has an Error Code and Reason Phrase, though I'm not sure why?  It'd be easy enough to specify an Error Code if we wanted to, so maybe that's enough?

the length of the Message Payload, the receiver MUST close the session with a
`PROTOCOL_VIOLATION`.

## Request ID {#request-id}

Request ID is included in request messages and is used to identify
requests across messages. For example, Joining Fetch references
the Request ID of a SUBSCRIBE.

The client generates even numbered Request IDs, starting at 0, and the
server generates odd numbered Request IDs, starting at 1.  Each
endpoint increments its Request ID by 2 for each new request.

Each SUBSCRIBE, PUBLISH, FETCH, SUBSCRIBE_NAMESPACE, PUBLISH_NAMESPACE,
REQUEST_UPDATE, and TRACK_STATUS message consumes a Request ID. Only
request messages include a Request ID; response messages do not, since
they are sent on the same bidirectional stream.

If an endpoint receives a Request ID where the least significant bit is
incorrect for the sender, or a duplicate Request ID, it MUST close the
session with `INVALID_REQUEST_ID`.

## Required Request ID {#required-request-id}

Every request message includes a Required Request ID Delta field, which is a delta
subtracted from Request ID to determine the Required Request ID. The receiver

> MUST deliver the referenced request to the application before delivering this request.  Required Request ID MUST have the same parity (least significant bit) as Request ID

Reviewer: This should be half-delta or delta-count. Odd numbers for delta never make sense.

might add clarity to explain how to compute required request id from the delta.

the length of the Message Payload, the receiver MUST close the session with a
`PROTOCOL_VIOLATION`.

## Request ID {#request-id}

Request ID is included in request messages and is used to identify
requests across messages. For example, Joining Fetch references
the Request ID of a SUBSCRIBE.

The client generates even numbered Request IDs, starting at 0, and the
server generates odd numbered Request IDs, starting at 1.  Each
endpoint increments its Request ID by 2 for each new request.

Each SUBSCRIBE, PUBLISH, FETCH, SUBSCRIBE_NAMESPACE, PUBLISH_NAMESPACE,
REQUEST_UPDATE, and TRACK_STATUS message consumes a Request ID. Only
request messages include a Request ID; response messages do not, since

> they are sent on the same bidirectional stream.

Reviewer 2: ```suggestion
they are sent on the same bidirectional stream as the request.
```

If an endpoint receives a Request ID where the least significant bit is
incorrect for the sender, or a duplicate Request ID, it MUST close the
session with `INVALID_REQUEST_ID`.

## Required Request ID {#required-request-id}

Every request message includes a Required Request ID Delta field, which is
multiplied by two and subtracted from Request ID to determine the

> Required Request ID. The receiver MUST deliver the referenced request to the

Reviewer 2: I am totally fine for this to be not delta encoded. It adds no value, i feel. Am i missing something ?

## SUBSCRIBE_OK {#message-subscribe-ok}

A publisher sends a SUBSCRIBE_OK as the first response message on the

> bidi stream for successful subscriptions.

Reviewer: Do we need a section that explains that for a given request type, the first message on the response stream must be either REQUEST_ERROR or the request-specific ok message, and anything else is PROTOCOL_VIOLATION?

SUBSCRIBE: SUBSCRIBE_OK
FETCH: FETCH_OK
PUBLISH: PUBLISH_OK
TRACK_STATUS: REQUEST_OK
PUBLISH_NS: REQUEST_OK
SUBSCRIBE_NS: REQUEST_OK

Author: Maybe, though I might be inclined to do that later.  There might be other text about Requests to reorg into its own section.

## PUBLISH_OK {#message-publish-ok}

The subscriber sends a PUBLISH_OK as the first response message on the
bidi stream to acknowledge the successful authorization and acceptance of a
PUBLISH message, and establish a subscription.

~~~
PUBLISH_OK Message {
  Type (i) = 0x1E,
  Length (16),
  Number of Parameters (i),
  Parameters (..) ...,
}
~~~
{: #moq-transport-publish-ok format title="MOQT PUBLISH_OK Message"}

* Parameters: The parameters are defined in {{version-specific-params}}.

TODO: A similar section to SUBSCRIBE about how the publisher handles a
filter that is entirely behind Largest Object or is otherwise invalid.

## PUBLISH_DONE {#message-publish-done}

A publisher sends a `PUBLISH_DONE` message as the final message on the
subscription's bidi stream to indicate it is done publishing Objects for that

> subscription.  The Status Code indicates why the subscription

Author: ```suggestion
A publisher sends a `PUBLISH_DONE` message as the final message before
closing the subscription's bidi stream to indicate it is done publishing Objects
for that subscription.  The Status Code indicates why the subscription
```

~~~
{: #moq-transport-fetch-format title="MOQT FETCH Message"}

* Fetch ID: An ID used to identify the unidirectional stream containing

>   Objects requested by the Fetch. The ID MUST be unique within a Session.

Reviewer: I get that subscribe ID and fetch ID only need to be unique.  Should we offer a recommendation that implementations do anything in particular?

Author: ```suggestion
  Objects requested by the Fetch. The ID MUST be unique within a Session.
  To ensure Fetch IDs are not reused, implementations SHOULD issue them
  sequentially.
```

Streams aside from the control stream MAY be canceled due to congestion
or other reasons by either the publisher or subscriber. Early termination of a
unidirectional stream does not affect the MoQ application state, and therefore has
no effect on outstanding subscriptions. Termination of a bidi request stream
terminates the Subscription, Fetch, Track Status, Publish Namespace, or Subscribe Namespace
request. Publishers SHOULD NOT RESET the bidi response stream, and instead SHOULD send

> the appropriate final message indicating the response is complete and FIN the bidi stream.

Reviewer: 1) subscribers and publisher both respond to requests
2) what i've already sent request_ok before cancellation, are you saying I should send request_error?  How is that better than RESET_STREAM + Cancelled?

Streams aside from the control stream MAY be canceled due to congestion
or other reasons by either the publisher or subscriber. Early termination of a
unidirectional stream does not affect the MoQ application state, and therefore has
no effect on outstanding subscriptions. Termination of a bidi request stream
terminates the Subscription, Fetch, Track Status, Publish Namespace, or Subscribe Namespace
request. Publishers SHOULD NOT RESET the bidi response stream, and instead SHOULD send

> the appropriate final message indicating the response is complete and FIN the bidi stream.

Author: ```suggestion
request. When possible, Publishers SHOULD send a SUBSCRIBE_DONE when terminating a
subscription instead of abruptly terminating the stream.
```

not limited to:

* An Object in an open Subgroup exceeding its Delivery Timeout

> * Early termination of subscription due to stream cancellation

Reviewer: ```suggestion
* Early termination of subscription due to request cancellation
```

: An implementation specific error.

CANCELLED (0x1):
: The subscriber requested cancellation via STOP_SENDING, or the publisher
  ended the subscription, in which case PUBLISH_DONE ({{message-publish-done}})

>   will have a more detailed status code.

Reviewer: This existing text doesn't cover fetch data streams.

: An implementation specific error.

CANCELLED (0x1):
: The subscriber requested cancellation via STOP_SENDING, or the publisher
  ended the subscription, in which case PUBLISH_DONE ({{message-publish-done}})

>   will have a more detailed status code.

Author: ```suggestion
: The subscriber or publisher cancelled the Request. For Subscriptions,
  PUBLISH_DONE ({{message-publish-done}}) will have a more detailed status code.
```
